### ISSUE TITLE
[BUG] npx assumes /tmp has execute permission

### ISSUE THREAD

**Issue by therealgilles** [npm/cli#5139]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

I just updated npm today: `npm install -g npm@8.13.2`

Everything was working fine with the previous version. Now when I try to run npx, I get:
```SHELL
sh: 1: /tmp/npx6574749239.sh: Permission denied
```

npx (or something underneath) seems to assume that /tmp has execute permissions, which is known to be unsafe.

### Expected Behavior

npx should work even though /tmp does not have execute permissions.

### Steps To Reproduce

1. Running on:
```
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

2. Using npm 8.3.2
3. Run `npx`

### Environment

- npm: v8.3.2
- Node.js: v16.15.1
- OS Name: Ubuntu 20.04.4 LTS (Focal Fossa)
- System Model Name: Linux box
- npm config:
```ini
; node bin location = /home/.../.nvm/versions/node/v16.15.1/bin/node
; node version = v16.15.1
; npm local prefix = /home/...
; npm version = 8.13.2
; cwd = ...
; HOME = /home/...
; Run `npm config ls -l` to show all defaults.
```

**Comment by ljharb** [comment]
Why would that be unsafe? How could npx work if it can't execute things in a temp dir?

**Comment by therealgilles** [comment]
Hi @ljharb, we can discuss the pros/cons of execute permissions on the /tmp dir but it may not be useful. I know it's used as a hardening/security technique on a number of linux server nodes, especially when running a web server. Here is an [example](https://docs.plesk.com/en-US/12.5/advanced-administration-guide-linux/enhancing-security.68755/#o68757) with the Plex server.

Now the more important points in my mind are:
- This seems to be a new behavior as I haven't encountered the problem before. I could be wrong about this, feel free to correct me.
- When I try `npm config set tmp`, I get the following message:
```
npm WARN config tmp This setting is no longer used.  npm stores temporary files in a special
npm WARN config location in the cache, and they are managed by
npm WARN config     [`cacache`](http://npm.im/cacache).
```
which seems contradictory with using the /tmp directory.
- I don't mind having to set TMPDIR to give npx a location to execute scripts. It would be best if on failure, npx could output a more meaningful error message.

**Comment by wraithgar** [comment]
npx executes scripts out of a tmp subdirectory that it creates inside of the configured `cache` directory.  I suspect that config value is somehow being set to `/` which is why npx is trying to use `/tmp`.  npm does not try to use `/tmp` as a rule, the default is to use `~/.npm/tmp` on posix, and `%LocalAppData%\npm-cache\` on windows.

**Comment by wraithgar** [comment]
Looks like I was mistaken, and npm does use `os.tmpdir` to run scripts themselves https://github.com/npm/run-script/blob/8136ac01048ea16b5f1fec321ee2c73e2c67a181/lib/make-spawn-args.js#L75

@nlf is aware of this and is working to remove the tempfile approach altogether.

**Comment by nlf** [comment]
today's release of npm 8.14.0 changed the strategy for the temp files slightly so that the execute permission is no longer required. in the future i hope to remove the temp files entirely, but in the mean time removing the requirement of execute permissions should have you unblocked.

go ahead and reopen this issue if you continue to have problems after updating.

**Comment by therealgilles** [comment]
Thanks all for looking into this and providing a fix so quickly!

**Comment by nikky78** [comment]
I have the same issue:

npm --version
     8.15.0

sudo npx playwright install-deps
     Need to install the following packages:
       playwright@1.27.1
     Ok to proceed? (y) y
     /tmp/npx-e3398e4e.sh: 1: playwright: Permission denied

**Comment by chris-monahan** [comment]
Also getting this, despite the fix apparently being applied on 8.14.0

This is on NPM 8.15.0 and Ubuntu 20.04.5 LTS

```
$ npx webpack
/tmp/npx-f1e4d967.sh: 1: webpack: Permission denied
```

Weirdly `npx webpack-cli` works fine, which I can use as a workaround, but from what I see in this issue here this shouldn't be happening.
